juju/juju
# Migrate local charm no resource

## Description of change

If a model deployed a local charm that has resources defined, but did not define any resources for that application, the model would fail to migrate. Also a local charm with local resources would migrate one time, but fail on subsequent attempts.

This branch fixes both of those problems. Firstly, resource definitions for an application where there is no actual resource are now handled by the model import code in the state package, rather in the binary upload process where it attempted to push empty placeholders. Also validation of the resource fingerprint was loosened so it doesn't try to validate an empty string.

## QA steps

Long... firstly make sure you have juju-ci-tools charms repo available, as I'm using their local charm. The resource itself is a simple index.html file that contains a string that you can validate came from the resource.

Bootstrap a controller and add two models, one for the local charm with a resource, and one with a local charm without a resource.
```
juju bootstrap lxd source
juju add-model resource
juju deploy ~/canonical/juju-ci-tools/repository/charms/simple-resource-http --resource index=~/sandbox/index.html
juju add-model no-resource
juju deploy ~/canonical/juju-ci-tools/repository/charms/simple-resource-http
```
Ensure that the resource is being used by using a web browser to look at the ip address of the machine that the app is deployed to.

Wait for the models to be fully deployed then bootstrap another controller to migrate the models to.
```
juju bootstrap lxd target
juju switch source
juju migrate resource target
juju migrate no-resource target
```
Wait for the migration to complete then ensure they are in the target controller. Add a unit to the resource model to make sure the resource is available. 
```
juju models -c target
juju add-unit -m target:resource simple-resource-http
```
Ensure it is used by looking at the IP address of machine-1 from the resource model.
Now migrate the models back

```
juju switch target
juju migrate resource source
juju migrate no-resource source
```
Wait for the migration to complete, then add another unit to the resource model and check again.
```
juju add-unit -m source:resource simple-resource-http
```

## Documentation changes

No docs needed, just fixing migration bugs.

## Bug reference

https://bugs.launchpad.net/juju/+bug/1692610
https://bugs.launchpad.net/juju/+bug/1692646
